samcli/hook_packages/terraform/hooks/prepare.py

```diff
@@ -0,0 +1,379 @@
+    return cfn_properties
+
+
+def _get_property_extractor(property_name: str) -> PropertyBuilder:
+    """
+    Returns a PropertyBuilder function to extract the given property from a dict
+
+    Parameters
+    ----------
+    property_name: str
+        The name of the property to extract
+
+    Returns
+    -------
+    PropertyBuilder
+        function that takes in a dict and extracts the given property name from it
+    """
+    return lambda properties: properties.get(property_name)
+
+
+def _build_lambda_function_environment_property(tf_properties: dict) -> Optional[dict]:
+    """
+    Builds the Environment property of a CloudFormation AWS Lambda Function out of the
+    properties of the equivalent terraform resource
+
+    Parameters
+    ----------
+    tf_properties: dict
+        Properties of the terraform AWS Lambda function resource
+
+    Returns
+    -------
+    dict
+        The built Environment property of a CloudFormation AWS Lambda Function resource
+    """
+    environment = tf_properties.get("environment")
+    if not environment:
+        return None
+
+    for env in environment:
+        variables = env.get("variables")
+        if variables:
+            return {"Variables": variables}
+
+    # no variables
+    return None
+
+
+def _build_lambda_function_code_property(tf_properties: dict) -> Any:
+    """
+    Builds the Code property of a CloudFormation AWS Lambda Function out of the
+    properties of the equivalent terraform resource
+
+    Parameters
+    ----------
+    tf_properties: dict
+        Properties of the terraform AWS Lambda function resource
+
+    Returns
+    -------
+    dict
+        The built Code property of a CloudFormation AWS Lambda Function resource
+    """
+    filename = tf_properties.get("filename")
+    if filename:
+        return filename
+
+    code = {}
+    tf_cfn_prop_names = [("s3_bucket", "S3Bucket"), ("s3_key", "S3Key")]
+    for tf_prop_name, cfn_prop_name in tf_cfn_prop_names:
+        tf_prop_value = tf_properties.get(tf_prop_name)
+        if tf_prop_value is not None:
+            code[cfn_prop_name] = tf_prop_value
+    return code
+
+
+AWS_LAMBDA_FUNCTION_PROPERTY_BUILDER_MAPPING: PropertyBuilderMapping = {
+    "FunctionName": _get_property_extractor("function_name"),
+    "Architectures": _get_property_extractor("architectures"),
+    "Environment": _build_lambda_function_environment_property,
+    "Code": _build_lambda_function_code_property,
+    "Handler": _get_property_extractor("handler"),
+    "PackageType": _get_property_extractor("package_type"),
+    "Runtime": _get_property_extractor("runtime"),
+    "Layers": _get_property_extractor("layers"),
+    "Timeout": _get_property_extractor("timeout"),
+}
+
+RESOURCE_TRANSLATOR_MAPPING: Dict[str, ResourceTranslator] = {
+    TF_AWS_LAMBDA_FUNCTION: ResourceTranslator(CFN_AWS_LAMBDA_FUNCTION, AWS_LAMBDA_FUNCTION_PROPERTY_BUILDER_MAPPING),
+}
+
+
+def _build_cfn_logical_id(tf_address: str) -> str:
```

Comment:
Let's add a comment to mark this function as critical, Also add a reference to a cfn doc on logical ids if you can.

changing IDs in the future can cause unintentional re-deployments.

Reply:
we will not deploy the generated Metadata file (CFN template) .. for IaCs support, we are not going to support `sam deploy` and `sam package` commands. Also, in case of `sam sync` where we need to deploy the underling application we will use the `terraform deploy` to deploy the application (in other IaCs, we are going to use the IaC tools to deploy, we will not deploy our own generated templates)